Makefile

```diff
@@ -1,111 +1,115 @@
-GCR_URL = us.gcr.io/vcm-ml
-DOCKERFILE ?= docker/Dockerfile
-COMPILED_TAG_NAME ?= latest
-ENVIRONMENT_TAG_NAME ?= latest
-COMPILE_OPTION ?=
-COMPILE_TARGET ?= fv3gfs-compiled
-BUILD_ARGS ?=
+# setup (use XXX=<value> make <target> to override)
+GCR_URL                 ?= us.gcr.io/vcm-ml
+DOCKERFILE              ?= docker/Dockerfile
+COMPILED_TAG_NAME       ?= latest
+ENVIRONMENT_TAG_NAME    ?= latest
+COMPILE_OPTION          ?=
+COMPILE_TARGET          ?= fv3gfs-compiled
+BUILD_ARGS              ?=
 BUILD_FROM_INTERMEDIATE ?= n
-ENVIRONMENT_TARGET ?= fv3gfs-environment
-COMPILED_IMAGE ?= $(GCR_URL)/$(COMPILE_TARGET):$(COMPILED_TAG_NAME)
-SERIALIZE_IMAGE ?= $(GCR_URL)/$(COMPILE_TARGET):$(COMPILED_TAG_NAME)-serialize
-ENVIRONMENT_IMAGE=$(GCR_URL)/$(ENVIRONMENT_TARGET):$(ENVIRONMENT_TAG_NAME)
-IMAGE ?= $(ENVIRONMENT_IMAGE)
-
-FMS_IMAGE = $(GCR_URL)/fms-build
-ESMF_IMAGE = $(GCR_URL)/esmf-build
-SERIALBOX_IMAGE = $(GCR_URL)/serialbox-build
-
-MOUNTS?=-v $(shell pwd)/FV3:/FV3 \
-	-v $(shell pwd)/FV3/conf/configure.fv3.gnu_docker:/FV3/conf/configure.fv3
-
-MOUNTS_SERIALIZE?=-v $(shell pwd)/FV3:/FV3/original
-
-EXPERIMENT ?= new
-RUNDIR_CONTAINER=/rundir
-RUNDIR_HOST=$(shell pwd)/rundir
+ENVIRONMENT_TARGET      ?= fv3gfs-environment
 
+# image names (use XXX_IMAGE=<name> make <target> to override)
+COMPILED_IMAGE     ?= $(GCR_URL)/$(COMPILE_TARGET):$(COMPILED_TAG_NAME)
+SERIALIZE_IMAGE    ?= $(GCR_URL)/$(COMPILE_TARGET):$(COMPILED_TAG_NAME)-serialize
+ENVIRONMENT_IMAGE  ?= $(GCR_URL)/$(ENVIRONMENT_TARGET):$(ENVIRONMENT_TAG_NAME)
+FMS_IMAGE          ?= $(GCR_URL)/fms-build
+ESMF_IMAGE         ?= $(GCR_URL)/esmf-build
+SERIALBOX_IMAGE    ?= $(GCR_URL)/serialbox-build
 
+# used to shorten build times in CircleCI
 ifeq ($(BUILD_FROM_INTERMEDIATE),y)
 	BUILD_ARGS += --build-arg FMS_IMAGE=$(FMS_IMAGE) --build-arg ESMF_IMAGE=$(ESMF_IMAGE) --build-arg SERIALBOX_IMAGE=$(SERIALBOX_IMAGE)
 endif
 
-build: build_compiled
+.PHONY: help
+help:
+	@awk 'BEGIN {FS = ":.*?## "} /^[a-zA-Z_-]+:.*?## / {printf "\033[36m%-30s\033[0m %s\n", $$1, $$2}' $(MAKEFILE_LIST)
 
-build_environment:
+.PHONY: build
+build: build_compiled  ## build default container image (production)
+
+.PHONY: build_environment
+build_environment:  ## build environment container image
 	docker build -f $(DOCKERFILE) -t $(ENVIRONMENT_IMAGE) \
 	--target $(ENVIRONMENT_TARGET) .
 
-build_compiled:
+.PHONY: build_compiled
+build_compiled:  ## build production container image
 	docker build \
 		--build-arg compile_option=$(COMPILE_OPTION) \
 		$(BUILD_ARGS) \
 		-f $(DOCKERFILE) \
 		-t $(COMPILED_IMAGE) \
 		--target $(COMPILE_TARGET) .
 
-build_serialize:
-	BUILD_ARGS="$(BUILD_ARGS) --build-arg serialize=true" COMPILED_IMAGE=$(SERIALIZE_IMAGE) $(MAKE) build_compiled
+.PHONY: build_serialize_check_md5
+build_serialize_check_md5:
+	BUILD_ARGS="$(BUILD_ARGS) --build-arg serialize=true" \
+	COMPILED_IMAGE=$(SERIALIZE_IMAGE)-check-md5 \
+	$(MAKE) build_compiled
+
+.PHONY: build_serialize
```

Comment:
I don't think 'build_serialize' should have GT4PY_DEV on -- although that is the main mode we are generating data in, these are two separate concepts that I think should be named as such. The default for 'make build_compiled' does not have GT4PY_DEV on. You should be able to compare the output from 'build_compiled' and 'build_serialize,' and I think adding the GT4PY_DEV flag should be named differently for both as well.

Reply:
<strike>Can I suggest naming `build_serialize_gt4py_dev` and `build_serialize`? This captures the fact that `build_compiled` and `build_serialize` differ by adding serialization, and `build_serialize` and `build_serialize_gt4py_dev` differ by adding the gt4py_dev flag. Currently it isn't clear to me that `build_serialize` would have a different COMPILE_OPTION than `build_compiled`.</strike>

(Added strikethrough above when I realized that this is what the old convention was. I think the old convention was clearer.)

Is there use for a `build_compiled_gt4py_dev` option?

Reply:
BTW, `enter_serialize` does not correspond to any compilation of the code, it bind-mounts the local sources over top of the compiled code and depends on you to build the model in the way you want while you're in the docker container. So you can't really say GT4PY_DEV is or isn't set for `enter_serialize`.

Reply:
That was exactly the point, which makes `enter_serialize` dangerous, since the user has to `GT4PY_DEV=Y make <target>` for developing serialization. We should find a way of making that clear or document it.

Reply:
Regarding the naming, I'll give it a final shot of trying to convince @rheacangeo before giving up.

Reply:
Name change undone. Will discuss further and potentially be the topic of another PR.

Reply:
In the current code build_compiled_gt4py_dev is the main pathway used to generate serialized data for fv3core regresssion tests, basically because we do things with GT4PY_DEV set that allow our regressions to pass without doing an allReduce, or rewriting the same math with three different orderings.
I see your point Oli, and simple names for the things we use the most makes sense to me too. I am not heels dug in on this, I think both the ways have points they clear up and points they make more confusing, but hopefully not too hard to deal with in either case. What should we name it when we turn GT4PY_DEV on for the non-serialized version?